index.js

```diff
@@ -78,420 +65,156 @@ class ServerlessAppsyncPlugin {
+  getApiKeyResources(config) {
+    if (config.authenticationType !== 'API_KEY') {
+      return {};
     }
-
-    return this.provider
-      .request('AppSync', 'updateGraphqlApi', config)
-      .then((data) => {
-        this.serverless.cli.log(`GraphQL API ID: ${data.graphqlApi.apiId}`);
-        this.serverless.cli.log(`GraphQL Endpoint: ${data.graphqlApi.uris.GRAPHQL}`);
-        // NOTE: storing the config in the appSync object
-        this.serverless.service.custom.appSync.awsResult = data;
-
-        if (resolvedConfig.authenticationType === 'API_KEY') {
-          this.updateAPIKey();
-        }
-      });
+    return {
+      GraphQlApiKeyDefault: {
+        Type: 'AWS::AppSync::ApiKey',
+        Properties: {
+          ApiId: { 'Fn::GetAtt': ['GraphQlApi', 'ApiId'] },
+          Description: 'serverless-appsync-plugin: Default',
+          Expires: Math.floor(Date.now() / 1000) + (365 * 24 * 60 * 60),
+        },
+      },
+    };
   }
 
-  attachDataSources() {
-    this.serverless.cli.log('Attaching data sources...');
-    const resolvedConfig = this.serverless.service.custom.appSync
-      .resolvedConfig;
-    const awsResult = this.serverless.service.custom.appSync.awsResult;
-
-    // eslint-disable-next-line arrow-body-style
-    const datasourceParams = resolvedConfig.dataSources.map((ds) => {
-      let config;
-      switch (ds.type) {
-        case 'AWS_LAMBDA':
-          config = {
-            lambdaConfig: {
-              lambdaFunctionArn: ds.config.lambdaFunctionArn,
-            },
-          };
-          break;
-        case 'AMAZON_DYNAMODB':
-          config = {
-            dynamodbConfig: {
-              awsRegion: resolvedConfig.region,
-              tableName: ds.config.tableName,
-            },
-          };
-          if (ds.config.useCallerCredentials) {
-            Object.assign(config, {
-              useCallerCredentials: ds.config.useCallerCredentials,
-            });
-          }
-          break;
-        case 'AMAZON_ELASTICSEARCH':
-          config = {
-            elasticsearchConfig: {
-              awsRegion: resolvedConfig.region,
-              endpoint: ds.config.endpoint,
-            },
-          };
-          break;
-        case 'NONE':
-          config = {};
-          break;
-        default:
-          this.serverless.cli.log('Data Source Type not supported', ds.type);
-      }
-      const dataSource = {
-        apiId: awsResult.graphqlApi.apiId,
-        name: ds.name,
-        type: ds.type,
-        description: ds.description,
-        serviceRoleArn: ds.config.serviceRoleArn,
+  getDataSourceResources(config) {
+    return config.dataSources.reduce((acc, ds) => {
+      const resource = {
+        Type: 'AWS::AppSync::DataSource',
+        Properties: {
+          ApiId: { 'Fn::GetAtt': ['GraphQlApi', 'ApiId'] },
+          Name: ds.name,
+          Description: ds.description,
+          Type: ds.type,
+          ServiceRoleArn: ds.config.serviceRoleArn,
```

Comment:
ServiceRoleArn is not required for Local resolvers (those of type "NONE"). This causes the following error:
`TypeError: Cannot read property 'serviceRoleArn' of undefined`. 

```
      - type: NONE
        name: Local
        description: "Local data source for mapping fields."
```